### PR TITLE
Enable custom CMS page templates,  title rendering is now optional

### PR DESCRIPTION
**Pull request by maximlion**
1. Simple CMS can now have multiple templates which are chosen in page admin.
2. Title can be visible/invisible depending on what the merchant wants for the specific page.
3. Added Sidebar templates to Simple CMS.

- Sidebar template example:
screenshot from 2018-08-30 22-29-07

- Title not being rendered example:
screenshot from 2018-08-30 22-29-35

Refs #1246

### CODE REVIEW

File: doc/changelog.rst [hunk 1]
```diff
@@ -47,10 +47,12 @@ Front
 - Bump template helpers caches when shop products and manufacturers are saved
 - Add edit in admin button to toolbar to enable editing the current object in admin page
 - Replace CDN styles with local packages
+- Title visibility in CMS pages is now optional
 
 General
 ~~~~~~~
 
+- Enable custom templates for Simple_CMS
```
Comment (chessbr) [inline doc/changelog.rst]: create a `Simple CMS` section for this

File: shuup/admin/modules/content/forms.py [hunk 2]
```diff
@@ -274,7 +274,9 @@ def _handle_simple_cms_save(self):
                     title=title,
                     content=rendered_content,
                     visible_in_menu=False,
+                    render_title=True,
```
Comment (chessbr) [inline shuup/admin/modules/content/forms.py]: you don't need this, it is `True` by default

File: shuup/simple_cms/admin_module/views.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 from django.conf import settings
 from django.core.exceptions import ValidationError
-from django.forms import DateTimeField
+from django.forms import DateTimeField, ChoiceField
```
Comment (chessbr) [inline shuup/simple_cms/admin_module/views.py]: not correctly sorted

File: shuup/simple_cms/admin_module/views.py [hunk 4]
```diff
@@ -26,6 +26,8 @@
 class PageForm(MultiLanguageModelForm):
     available_from = DateTimeField(label=_("Available from"), required=False, localize=True)
     available_to = DateTimeField(label=_("Available to"), required=False, localize=True)
+    template_path = ChoiceField(label=_("Choose a template"),
```
Comment (chessbr) [inline shuup/simple_cms/admin_module/views.py]: remember: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/

```py
template_path = ChoiceField(
    label=_("Choose a template"),
    choices=settings.SHUUP_SIMPLE_CMS_TEMPLATES,
    required=False
)
```

File: shuup/simple_cms/migrations/0010_auto_20180830_2312.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
```
Comment (chessbr) [inline shuup/simple_cms/migrations/0010_auto_20180830_2312.py]: put a name on this migration file, do not use the auto generated file name

File: shuup/simple_cms/models.py [hunk 6]
```diff
@@ -111,8 +111,16 @@ class Page(MPTTModel, TranslatableModel):
         ),
         content=models.TextField(verbose_name=_('content'), help_text=_(
             "The page content. This is the text that is displayed when customers click on your page link."
+            "You can leave this empty and add all page content through placeholder editor in shop front."
+            "To edit the style of the page you can use the Snippet plugin which is in shop front editor."
         )),
     )
+    template_path = models.CharField(max_length=500, verbose_name=_("Template path"),
```
Comment (chessbr) [inline shuup/simple_cms/models.py]: code styling..

File: shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja [hunk 7]
```diff
@@ -0,0 +1,44 @@
+{% extends "shuup/front/base.jinja" %}
+
+{% block title %}{{ page.title }}{% endblock %}
+{% block content_title %}
+    {% if page.render_title %}
+            {{ page.title }}
```
Comment (chessbr) [inline shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja]: wrong identation

File: shuup_workbench/settings/base_settings.py [hunk 8]
```diff
@@ -246,6 +246,13 @@
     from .utils import get_disabled_migrations
     MIGRATION_MODULES = get_disabled_migrations()
 
+SHUUP_SIMPLE_CMS_DEFAULT_TEMPLATE = "shuup/simple_cms/page.jinja"
+# Remember to restart the server after any change to the list below
```
Comment (chessbr) [inline shuup_workbench/settings/base_settings.py]: remove this comment, this is required for any change in settings

File: shuup_workbench/settings/base_settings.py [hunk 9]
```diff
@@ -246,6 +246,13 @@
...
+SHUUP_SIMPLE_CMS_TEMPLATES = [
```
Comment (chessbr) [inline shuup_workbench/settings/base_settings.py]: I guess using provides would be a way better solution for this

File: shuup/simple_cms/views.py [hunk 10]
```diff
@@ -46,6 +47,11 @@ def get(self, request, *args, **kwargs):
         context = self.get_context_data(object=self.object)
         return self.render_to_response(context)
 
+    def get_template_names(self):
+        object = self.get_object()
+
```
Comment (chessbr) [inline shuup/simple_cms/views.py]: remove blank line

File: shuup/simple_cms/admin_module/views.py [hunk 11]
```diff
@@ -26,6 +27,14 @@
...
     available_from = DateTimeField(label=_("Available from"), required=False, localize=True)
     available_to = DateTimeField(label=_("Available to"), required=False, localize=True)
+    choices = []
+    for simple_cms_template in get_provide_objects("simple_cms_template"):
+        choices.append((simple_cms_template.path, simple_cms_template.name))
+    template_path = ChoiceField(
+        label=_("Choose a template"),
+        choices=choices,
+        required=False
+        )
```
Comment (chessbr) [inline shuup/simple_cms/admin_module/views.py]: make `)` back a column

File: shuup/simple_cms/admin_module/views.py [hunk 12]
```diff
@@ -38,8 +47,10 @@ class Meta:
             'identifier',
             'visible_in_menu',
             'parent',
+            'template_path',
```
Comment (chessbr) [inline shuup/simple_cms/admin_module/views.py]: replace `template_path` by `tamplate_name` a common used variable name to set templates

File: shuup/simple_cms/templates.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+# Copyright (c) 2012-2018, Shuup Inc. All rights reserved.
+#
+# This source code is licensed under the OSL-3.0 license found in the
+# LICENSE file in the root directory of this source tree.
+from django.utils.translation import ugettext as _
+
+
+class SimpleCMSTemplate(object):
+    name = _("Default")
+    path = "shuup/simple_cms/page.jinja"
```
Comment (chessbr) [inline shuup/simple_cms/templates.py]: use `template_path`

File: shuup_workbench/settings/base_settings.py [hunk 14]
```diff
@@ -246,6 +246,8 @@
     from .utils import get_disabled_migrations
     MIGRATION_MODULES = get_disabled_migrations()
 
+SHUUP_SIMPLE_CMS_DEFAULT_TEMPLATE = "shuup/simple_cms/page.jinja"
```
Comment (chessbr) [inline shuup_workbench/settings/base_settings.py]: move this to `simple_cms/settings.py` and put documentation

File: shuup_workbench/test_settings.py [hunk 15]
```diff
@@ -234,6 +234,8 @@
     from .settings.utils import get_disabled_migrations
     MIGRATION_MODULES = get_disabled_migrations()
 
+SHUUP_SIMPLE_CMS_DEFAULT_TEMPLATE = "shuup/simple_cms/page.jinja"
```
Comment (chessbr) [inline shuup_workbench/test_settings.py]: remove this from here

File: shuup/simple_cms/admin_module/views.py [hunk 16]
```diff
@@ -26,6 +27,14 @@
 class PageForm(MultiLanguageModelForm):
     available_from = DateTimeField(label=_("Available from"), required=False, localize=True)
     available_to = DateTimeField(label=_("Available to"), required=False, localize=True)
+    choices = []
+    for simple_cms_template in get_provide_objects("simple_cms_template"):
+        choices.append((simple_cms_template.path, simple_cms_template.name))
+    template_path = ChoiceField(
+        label=_("Choose a template"),
```
Comment (chessbr) [inline shuup/simple_cms/admin_module/views.py]: just "Template"

File: shuup/simple_cms/models.py [hunk 17]
```diff
@@ -111,8 +111,19 @@ class Page(MPTTModel, TranslatableModel):
         ),
         content=models.TextField(verbose_name=_('content'), help_text=_(
             "The page content. This is the text that is displayed when customers click on your page link."
+            "You can leave this empty and add all page content through placeholder editor in shop front."
+            "To edit the style of the page you can use the Snippet plugin which is in shop front editor."
         )),
     )
+    template_path = models.CharField(
```
Comment (chessbr) [inline shuup/simple_cms/models.py]: use TextField here, a super large CharField doesn't make sense

File: shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja [hunk 18]
```diff
@@ -0,0 +1,44 @@
+{% extends "shuup/front/base.jinja" %}
+
+{% block title %}{{ page.title }}{% endblock %}
+{% block content_title %}
+    {% if page.render_title %}
+        {{ page.title }}
+    {% endif %}
+    <div class="col-md-3 categories-nav" style="font-size: 15px">
```
Comment (chessbr) [inline shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja]: this just doesn't make sense to be used here.. this is content, not title

File: shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    <ol class="breadcrumb">
+        <li><a href="/">{% trans %}Home{% endtrans %}</a></li>
+        {% for ancestor in page.get_ancestors() if ancestor.is_visible() %}
+            <li><a href="/{{ ancestor.url }}">{{ ancestor.title }}</a></li>
+        {% endfor %}
+        <li class="active">{{ page.title }}</li>
+    </ol>
+{% endblock %}
+{% block content %}
+
```
Comment (chessbr) [inline shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja]: this template should looks like this:
```jinja
{% extends "shuup/front/base.jinja" %}

{% block title %}{{ page.title }}{% endblock %}
{% block content_title %}
    {% if page.render_title %}
        {{ page.title }}
    {% endif %}
{% endblock %}
{% block extrameta %}
{{ macros.render_metadata(page, page.title, page.get_html(), "page") }}
{% endblock %}

{% block breadcrumb %}
    <ol class="breadcrumb">
        <li><a href="/">{% trans %}Home{% endtrans %}</a></li>
        {% for ancestor in page.get_ancestors() if ancestor.is_visible() %}
            <li><a href="/{{ ancestor.url }}">{{ ancestor.title }}</a></li>
        {% endfor %}
        <li class="active">{{ page.title }}</li>
    </ol>
{% endblock %}
{% block content %}
    <div class="col-md-3 categories-nav">
        {{ category_macros.render_category_menu() }}
    </div>
    <div class="col-md-8">
        {{ page.get_html()|safe }}
        {% if page.list_children_on_page %}
            {# TODO: add pagination for children list #}
            {% for child in page.children.visible(request.shop).order_by("-available_from") %}
                <div class="row">
                    <a href="/{{ child.url }}"><h2>{{ child.title }}</h2></a>
                    {% if child.available_from and page.show_child_timestamps %}
                        <small>{{ child.available_from|datetime }}</small>
                    {% endif %}
                    <p>{{ child.get_html()|safe|striptags|truncate(256) }}</p>
                </div>
            {% endfor %}
        {% endif %}
        {% placeholder "cms_page" %}{% endplaceholder %}
    </div>
{% endblock %}
{% block extrajs %}
    {{ category_macros.render_extrajs() }}
{% endblock %}
```

Comment (maximlion) [reply shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja]: The title is `above` the sidebar, also title & sidebar aren't aligned.

screenshot from 2018-08-31 23-48-37

Comment (chessbr) [reply shuup/simple_cms/templates/shuup/simple_cms/custom_sidebar.jinja]: your solution doesn't work either, the sidebar styles are completely broken, you shouldn't have to apply element inline styles

**Pull request merged**